### PR TITLE
new feature - consul_exporter

### PR DESCRIPTION
**Pull request by pavloos**
- added consul_exporter

### CODE REVIEW

**Comment by pavloos** [comment]
bump

**Comment by pavloos** [comment]
Hello, Can this PR be looked into please ?

**Comment by liamjbennett** [comment]
This looks good to me. @bastelfreak any issues with me merging?

File: manifests/consul_exporter.pp [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+  $web_listen_address    = $::prometheus::params::consul_exporter_web_listen_address,
+  $web_telemetry_path    = $::prometheus::params::consul_exporter_web_telemetry_path,
+) inherits prometheus::params {
+  # Prometheus added a 'v' on the realease name at 0.3.0
+  if versioncmp ($version, '0.3.0') == -1 {
+    fail("I only support consul_exporter version '0.3.0' or higher")
+  }
+
+  $real_download_url = pick($download_url,"${download_url_base}/download/v${version}/${package_name}-${version}.${os}-${arch}.${download_extension}")
+  validate_bool($purge_config_dir)
```
Comment (bastelfreak) [inline manifests/consul_exporter.pp]: Hi @pjfbashton, could you replace those legacy functions with native puppet datatypes?

File: manifests/consul_exporter.pp [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+#  [*version*]
+#  The binary release version
+#
+#  [*web_listen_address*]
+#  Address to listen on for web interface and telemetry. (default ":9107")
+#
+#  [*web_telemetry_path*]
+#  Path under which to expose metrics. (default "/metrics")
+class prometheus::consul_exporter (
+  String           $arch                  = $::prometheus::params::arch,
```
Comment (bastelfreak) [inline manifests/consul_exporter.pp:87]: datatypes \o/

**Comment by bastelfreak** [comment]
@pjfbashton thanks for the datatypes. Can you add at least a little unit test that verifies if the catalog compiles?

**Comment by pavloos** [comment]
@bastelfreak Oh, those tests failed a while ago, have changed anything ? I think there is a problem with prometheus service - it does not start (at least on Centos 7) on the first run with default settings (just `include ::prometheus`). Is this known bug? 

I can try adding unit test although it is new to me, have you got an example I could take a look at?

**Comment by bastelfreak** [comment]
One example is at: https://github.com/voxpupuli/puppet-prometheus/blob/master/spec/classes/beanstalkd_exporter_spec.rb

I am not really sure why the acceptance tests fail randomly. I wasn't able to reproduce this locally. Sometimes they work, sometimes they don't :(

**Comment by pavloos** [comment]
Failed again on prometheus service not starting after first run. The same happens for me locally... 

```
[root@e789cf77067a ~]# cat /etc/redhat-release 
CentOS Linux release 7.2.1511 (Core) 
[root@e789cf77067a ~]# puppet -V
4.10.5

puppet module install puppet-archive -v 2.1.0
puppet module install puppetlabs-stdlib -v 4.22.0
puppet module install camptocamp-systemd -v 1.1.0
git clone --branch v3.0.0 https://github.com/voxpupuli/puppet-prometheus.git \
  && mv puppet-prometheus/ /etc/puppetlabs/code/modules/prometheus
echo "include prometheus" > p.pp

[root@e789cf77067a ~]# puppet apply --verbose p.pp
Info: Loading facts
Info: Loading facts
Info: Loading facts
Warning: /etc/puppetlabs/puppet/hiera.yaml: Use of 'hiera.yaml' version 3 is deprecated. It should be converted to version 5
   (in /etc/puppetlabs/puppet/hiera.yaml)
Notice: Compiled catalog for e789cf77067a.xps13.pjf in environment production in 0.45 seconds
Info: Applying configuration version '1510830353'
Notice: /Stage[main]/Prometheus::Install/Archive[/tmp/prometheus-1.5.2.tar.gz]/ensure: download archive from https://github.com/prometheus/prometheus/releases/download/v1.5.2/prometheus-1.5.2.linux-amd64.tar.gz to /tmp/prometheus-1.5.2.tar.gz and extracted in /opt with cleanup
Notice: /Stage[main]/Prometheus::Install/File[/opt/prometheus-1.5.2.linux-amd64/prometheus]/owner: owner changed 'pawel' to 'root'
Notice: /Stage[main]/Prometheus::Install/File[/opt/prometheus-1.5.2.linux-amd64/prometheus]/group: group changed 'pawel' to 'root'
Notice: /Stage[main]/Prometheus::Install/File[/opt/prometheus-1.5.2.linux-amd64/prometheus]/mode: mode changed '0755' to '0555'
Notice: /Stage[main]/Prometheus::Install/File[/usr/local/bin/prometheus]/ensure: created
Info: /Stage[main]/Prometheus::Install/File[/usr/local/bin/prometheus]: Scheduling refresh of Service[prometheus]
Notice: /Stage[main]/Prometheus::Install/Group[prometheus]/ensure: created
Notice: /Stage[main]/Prometheus::Install/User[prometheus]/ensure: created
Notice: /Stage[main]/Prometheus::Install/File[/var/lib/prometheus]/ensure: created
Notice: /Stage[main]/Prometheus::Install/File[/usr/local/share/prometheus]/ensure: created
Notice: /Stage[main]/Prometheus::Install/File[/usr/local/share/prometheus/consoles]/ensure: created
Info: /Stage[main]/Prometheus::Install/File[/usr/local/share/prometheus/consoles]: Scheduling refresh of Service[prometheus]
Notice: /Stage[main]/Prometheus::Install/File[/usr/local/share/prometheus/console_libraries]/ensure: created
Info: /Stage[main]/Prometheus::Install/File[/usr/local/share/prometheus/console_libraries]: Scheduling refresh of Service[prometheus]
Notice: /Stage[main]/Prometheus::Config/File[/etc/prometheus]/ensure: created
Info: /Stage[main]/Prometheus::Config/File[/etc/prometheus]: Scheduling refresh of Class[Prometheus::Run_service]
Notice: /Stage[main]/Prometheus::Config/File[prometheus.yaml]/ensure: defined content as '{md5}1463d85961bf7b73da62eaf7be1eaef9'
Info: /Stage[main]/Prometheus::Config/File[prometheus.yaml]: Scheduling refresh of Class[Prometheus::Service_reload]
Notice: /Stage[main]/Prometheus::Config/Systemd::Unit_file[prometheus.service]/File[/etc/systemd/system/prometheus.service]/ensure: defined content as '{md5}5ca91801aafbad71eec1c195d4d508dc'
Info: /Stage[main]/Prometheus::Config/Systemd::Unit_file[prometheus.service]/File[/etc/systemd/system/prometheus.service]: Scheduling refresh of Class[Systemd::Systemctl::Daemon_reload]
Info: Class[Systemd::Systemctl::Daemon_reload]: Scheduling refresh of Exec[systemctl-daemon-reload]
Notice: /Stage[main]/Systemd::Systemctl::Daemon_reload/Exec[systemctl-daemon-reload]: Triggered 'refresh' from 1 events
Info: Class[Prometheus::Run_service]: Scheduling refresh of Service[prometheus]
Notice: /Stage[main]/Prometheus::Run_service/Service[prometheus]/ensure: ensure changed 'stopped' to 'running'
Info: /Stage[main]/Prometheus::Run_service/Service[prometheus]: Unscheduling refresh on Service[prometheus]
Info: Class[Prometheus::Service_reload]: Scheduling refresh of Exec[prometheus-reload]
Notice: /Stage[main]/Prometheus::Service_reload/Exec[prometheus-reload]: Triggered 'refresh' from 1 events
Info: Creating state file /opt/puppetlabs/puppet/cache/state/state.yaml
Notice: Applied catalog in 19.75 seconds
[root@e789cf77067a ~]# systemctl status prometheus
● prometheus.service - Prometheus Monitoring framework
   Loaded: loaded (/etc/systemd/system/prometheus.service; enabled; vendor preset: disabled)
   Active: inactive (dead) since Thu 2017-11-16 11:06:13 UTC; 23s ago
  Process: 409 ExecReload=/bin/kill -HUP $MAINPID (code=exited, status=0/SUCCESS)
 Main PID: 369 (code=killed, signal=HUP)

Nov 16 11:06:13 e789cf77067a systemd[1]: Started Prometheus Monitoring framework.
Nov 16 11:06:13 e789cf77067a systemd[1]: Starting Prometheus Monitoring framework...
Nov 16 11:06:13 e789cf77067a systemd[1]: Reloaded Prometheus Monitoring framework.
```

**Comment by pavloos** [comment]
all green now @bastelfreak

**Comment by tuxmea** [comment]
looks good to me. thank you for the PR.

**Pull request merged**